JO/AA MAPL3 implementation into MIRACL

jono3030 opened the pull request:
Implement `MAPL3` written by @AAttarpour into `MIRACL`.

~~This is the module implementation. The workflow implementation will be done in a separate PR.~~

This is the PR for both the module *and* the workflow implementation.

This implementation introduces Pydantic concepts to `MIRACL` that should eventually be adopted by all modules.

jono3030 committed:
[feat] add Pydantic datamodel for MIRACL objects

This is part of a wider refactoring effort of MIRACL. Unfortunately we
can't completely refactor MIRACL from the ground up due to time
constraints. However, this module, MAPL3 (final title TBD) introduces
Pydantic classes to MIRACL. Subsequent modules will be implemented the
same way, also using the Pydantic datamodel. Whenever time allows,
modules that do not use Pydantic yet, should be refactored. This will
hopefully allow us to gradually refactor MIRACL over time. This commit
is the first step in this direction.

jono3030 committed:
[feat] add refactored parser fn

This refactored parser fn helps with generating cli parsers using
Pydantic objects.

jono3030 committed:
[feat] add Pydantic data objects for the generating patches script from MAPL3

These are the objects for the generating patches script that will be
added for the MAPL3 implementation.

jono3030 committed:
[feat] add first version of the parser for the MAPL3 module

jono3030 committed:
[fix] fix incorrect gui_group name

jono3030 committed:
[feat] add first WIP implementation of MAPL3's generate patch script

This is the first implmementation of the generate patch script into the
interface and the first actual implementation of the interface for the
MAPL3 module. It's a WIP but the implementation is getting there.

jono3030 committed:
[feat] add more attributes to MIRACL objects

Adding these extra attributes will greatly improve the workflow and
validity of the interface.

jono3030 committed:
[feat] add more objects for MAPL3

jono3030 committed:
[feat] implement objects into interface and finish implementation of generate patches script

jono3030 committed:
[feat] explicitely add booleans to Pydantic obj_default attribute

jono3030 committed:
[feat] complete generate patch and parallel preprocessing Pydantic objects

jono3030 committed:
[feat] add argparse action to Pydantic-to-argparse converter after updating BaseModel with explicit booleans for obj_default

jono3030 committed:
[fix] remove if ... __main__

jono3030 committed:
[feat] first version of parallel preprocessing implementation

jono3030 committed:
[feat] add parallel preprocessing script

jono3030 committed:
[feat] call preprocessing script and replace argparse in it

jono3030 committed:
[fix] remove code used for testing

jono3030 committed:
[feat] implement inference and change interface folder object class

Implements inference script but it's a WIP. There are several variables
that get passed to the inference script that are not working yet. Mostly
this is because of type differences. I did improve the way subfolders
are handled as Pydantic objects. Each subfolder is a separate instance
of the interface subfolder class now.

jono3030 committed:
[feat] finish first implementation of MAPL3 module

Way too many changes for one commit. That's not good practice. I blame
it on the time crunch and at still being at the office at 1am. The
module is implemented but has not yet been integrated into MIRACL's CLI.

jono3030 committed:
[feat] improve BaseModel

This should have been part of the previous commit. Again, it's past 1pm
on December 24th and I'm still at the office. I blame it on that.

jono3030 committed:
[fix] add path for imports

jono3030 committed:
[wip] add first version of cli implementation

This is not working yet because of a joblib related problem. There are
different version of joblib installed in the container that need to be
sorted out.

jono3030 committed:
[fix] resolve package conflict

Pinning joblib to version 1.3.2 solves the following error:
`ImportError: cannot import name 'parallel_config' from
'joblib.parallel'`.

jono3030 committed:
[fix] fix incorrect parsing in MIRACL

The changes in the MAPL3 interface should be considered a temporary
solution. There must be a better way to solve this than skipping the
first two args (`seg mapl3`) to the parser.

jono3030 committed:
[feat] rewrite logger for more finegrained logging output

jono3030 committed:
[feat] improve interface debugging msgs and add rudimentary module enable/disable functionality

jono3030 committed:
[fix] fix incorrect args forwarding in preprocessing parallel main()

jono3030 committed:
[feat] improve user info output for parallel processing

jono3030 committed:
[feat] improve user info output for generate patch and check for paths

jono3030 committed:
[feat] improve user info output for patch stacking

jono3030 committed:
[feat] improve user info output for parallel preprocessing and remove folder creation as its guaranteed in interface

jono3030 committed:
[fix] fix incorrect args forwarding for preprocessed folder

jono3030 committed:
[feat] improve user info output for inference

jono3030 committed:
[feat] improve user info output for skeletonization

jono3030 committed:
[feat/fix] improve user info output for patch stacking and fix incorrect variable assignment

jono3030 committed:
[feat] improve module enable/disable slightly

This needs to be done properly in the future. It's good enough for now
because of time constraints.

jono3030 committed:
[feat] add Pydantic version 2.5.3 to setup.py

[fix] fix missing equal sign in setup.py

Damn you double equal signs!

AAttarpour reviewed (changes requested):
The structure is fine; great job, but I have one idea. Under the mapl3 directory, we need to have three sub-folders: 1) model, which contains transformers and building blocks; 2) preprocessing, which contains generating patch scripts; and 3) postprocessing, which contains skeletonization and patch stacking scripts.

jono3030 committed:
[feat/fix] add path to folder containing percentages json file to inference and skeletonization scripts

jono3030 committed:
[feat] add 'mapl3' and 'seg' subfolders to results folder for MAPL3 module interface

jono3030 committed:
[feat] pass parser to main() instead of hardcoding it within main()

This will allow us to pass the workflow parser in the workflow
interface. The workflow parser will import and add the module interface
parser and hence will contain all args from it plus the args for the
additional modules in the workflow.

jono3030 committed:
[feat] add first, WIP implementation of MAPL3 workflow interface

The MAPL3 interface now returns a dictionary with the objects that are
required by the other modules in the workflow. It would probably make
sense to wrap the object generation in the main() fn of the MAPL3
interface into a class to be able to return all objects and then pick
and choose in the workflow instead of hardcoding the returned objects in
the dictionary. I'll make a note of this and implmement it during
refactoring once the time crunch is less and MAPL3 is stable for the
module and the workflow.

jono3030 committed:
[feat] restructure parser dict to have one source of truth that can be imported in the workflow parser also

jono3030 committed:
[feat/wip] add Clar Allen registration objects

This is not working yet. The const versions of the objects don't work in
the cli parser yet.

jono3030 committed:
[feat/wip] first implementation of MAPL3 workflow

This is not working yet. The flags that are being used in the workflow
differ from the flags that are being used in the module. This is
intentional but when importing the MAPL3 module interface into the
workflow, the module interface expects the module flags/args to create
the objects with their content. However, the workflow args will be
parsed on the cli instead and the interface will throw an error because
an expected flag/args is missing.

jono3030 committed:
[feat] major refactor of interface and processing methods for Pydantic objs

Ideally I would have liked to push this as several commits instead of
one big commit but time constraints won't allow that. This refactor was
needed to accomodate using the module as part of a encompassing
workflow. The interface can now be imported into the workflow and the
appropriate workflow flags will be used automatically. This will work
even better once the cli parsing of MIRACL has been refactored with
Typer.

Major changes:

1) The parser creator is now a class that consists of a method to create
the parser objects and a method to process the parsed cli arguments and
assign them automagically to their respective object/attr.

2) The arguments for the parser creation are now separated from the
parser scripts. I also added a dictionary with all Pydantic objects that
are used in the interface. This gives us a single source of truth that
can be imported/returned for processing. This is particularly useful for
creating workflows as we will have a separate workflow dictionary with
all objects needed for it.

3) The interface for the module is much shorter now with a lot less
duplication and variable assignment since the cli args are assigned
automatically. They attributes for the objects can now be referenced
directly using the objs dict.

4) All folders that are required for MAPL3's in-module workflow are now
created from a separate objects file. That way they can be handled
separately from the core MIRACL objects.

To do:

1) The interface should return two dicts, namely the MIRACL core objects
and MAPL3's folders.
2) The workflow has to be refactored to the same structure as the
module.
3) The dataclasses should be added to the ABC/abstractmethods in the
interface to avoid duplication.

jono3030 committed:
[feat] improve interface with dataclasses

jono3030 committed:
[feat] first working version of MAPL3 workflow

A few things are still WIP but this is the first version of the MAPL3
workflow that actually works. The main additions here are the main
scripts for the MAPL3 module and workflows as well as re-defining the
input to the interfaces. This allows us to now call the MAPL3 module
interface with whatever parser/object combination that we want, in our
case either the module parser or the workflow parser.

jono3030 committed:
[fix] uncomment erroniously commented section

jono3030 committed:
[feat] first implementation of the conversion script into MAPL3 workflow

jono3030 committed:
[wip/feat] add converted nifti file from conv to reg fn

The Path object should be assigned to the filepath attribute of a MIRACL
obj instead of passing just a variable. However, due to time
contstraints this will have to do for now.

jono3030 committed:
[feat] add type annotations and docstrings to workflow interface and main

[fix] correct type annotation

jono3030 committed:
[feat] add missing docstrings and type annotations to MAPL3 module interface and main fn

Also small improvements to parser descriptions of objects and added
extra validation checks to MAPL3 subfolders.

jono3030 committed:
[feat] add extra validation for folder and file paths in MAPL3 interface

[feat] improve folder and file path validation functions

jono3030 commented:
Since we are already combining the module and workflow implementations into one PR here, why not also add the docs updates? I added a label and I suggest that we start adding documentation soon as we proceed further with testing the implementation.

jono3030 committed:
[feat/fix] improve parser creation method

Features:
1) Add Enums
2) Add path checker
3) Clean up code by e.g. removing previous versions that were
uncommented
Fixes:
1) Fix parsing for module vs. worklfow flags. The method now checks
whether the module is called directly or as part of a workflow and uses
the respective flag.
2) Added missing error/exceptions checking

jono3030 committed:
[feat] formatting improvements

jono3030 committed:
[wip] more work on the MAPL3 implementation

Also some formatting improvements, turning off incorrect logs and some
temporary sys.exit() calls for debugging.

jono3030 committed:
[wip] more refactoring

New approach to Pydantic datamodel. Instead of constructing with a processor I construct with a serializer. Included Pydantic object to argparse serializer and a deserializer from argparse object back to Pydantic object to trigger validation. I also separated the enums and fns from the datamodel class.

jono3030 committed:
[wip] more experimentation with automating argparser groups

jono3030 committed:
[wip] first version of godlike refactoring

So psyched!!

jono3030 committed:
[feat] add latest version of generate_patch script

jono3030 committed:
[feat] create/update generate_patch objects

jono3030 committed:
[fix] correct some object names for clar_allen objects

jono3030 committed:
[fix] small fix to argparse serializer

jono3030 committed:
[feat] add generate_patch ENUM's for CliGroup

jono3030 committed:
[feat] several improvements to registry

- Add new object to template/registry to pass actual class object references for use in parser
- Add template for generate_patch method
- Experiments to improve argparser and link it to registry

jono3030 committed:
[feat] significant refactor of serializers

Reducing code duplication and making methods work with flow attributes
instead of modules only.

jono3030 committed:
[fix] fix missing type annotations and incorrect ENUMS import

jono3030 committed:
[fix/feat] use ENUMS instead of str and remove flag_map code for templates

jono3030 committed:
[feat] use ENUMS in registry

jono3030 committed:
[wip] continue workflow execution testing with registry and argparser

jono3030 committed:
[wip] test cli_group attribute for argparser groups

jono3030 committed:
[feat] clean up argparser builder and switch to ENUM use for module_type

jono3030 committed:
[feat] small type annotation fix for command runner

jono3030 committed:
[feat] first working version of overrides with object references

The override values for the registry.run() can now be referenced
directly from the object attribute values!! This was a pain to implement
and will definitely need refactoring:

1) The parsing/building of the flag/mappings dict should be the
responsibility of the serializer not the registry
2) The registry (?) should have a wrapper function to make referencing
in the overrides method more convenient i.e. simplifying this:
```
obj_class = reg.get_class("conversion")

attr_instance = getattr(obj_class, "tiff_folder")

test_flag = attr_instance.flow.get("mapl3", {}).get("cli_l_flag")

test_value = getattr(
    reg.get_class("workflow_connectors"),
"mapl3_workflow_raw_tiff_folder"
).content

print(f"FLAG: {test_flag}")
print(f"VALUE: {test_value}")

print("")
reg.run(
    "conversion",
    overrides={
        f"--{test_flag}": test_value,
    },
)
print("")
```

jono3030 committed:
[fix] attr id should be of type UUID not str

jono3030 committed:
[fix] fix incorrect id for CLARITY Allen obj

jono3030 committed:
[wip] model dump tests for serializer improvements

Deserialization should use UUID for matching back to and updating object
attributes instead of name. Possibly using cli dest by assigning UUID to
it.

jono3030 committed:
[wip] idea for override referencing

jono3030 committed:
[fix] make flow definition more explicit in datamodel

cli_s_flag and cli_l_flag are required explicitly now.

jono3030 committed:
[feat/fix] improve metavar usage and fix type annotations and docstrings

A custom metavar can now be passed but if it isn't, metavar will be set
to obj.name. Previously obj.name would be used for all metavars in the
parser.

jono3030 committed:
[feat] make datamodel description for flow more explicit

jono3030 committed:
[fix] fix conditional logic for flag creator (serializer)

jono3030 committed:
[fix] further improvement to conditional logic and add missing argparser const option

jono3030 committed:
[feat] small streamlining of dict iteration

jono3030 committed:
[feat] major improvements in registry for overrides

Several type annotation fixes and added two new methods:

1. get_override_flag()
  - Get the CLI flag associated with a specific object in a registered
module
2. get_override_value()
  - Get the content value associated with a specific object in a
registered module

With these two methods, override values can be accessed directly with a
convenient wrapper that uses the registry directly to get the
information from the registered modules. E.g.:

reg.run(
    "conversion",
    overrides={
        f"--{reg.get_override_flag('conversion', 'tiff_folder')}":
reg.get_override_value(
            "workflow_connectors", "mapl3_workflow_raw_tiff_folder"
        ),
    },
)

Note: The flag override should return the formatted flag. Should be
fixed in the next commit.

jono3030 committed:
[fix] fix flag formatting for get_override_flag method

New syntax example:

reg.run(
    "conversion",
    overrides={
        reg.get_override_flag("conversion", "tiff_folder"):
reg.get_override_value(
            "workflow_connectors", "mapl3_workflow_raw_tiff_folder"
        ),
    },
)

Results in: Optional[Dict[str, object]] with prepended `--` for str.

jono3030 committed:
[feat] conv, reg and connector objs prepped for workflow

jono3030 committed:
[feat] custom runners for conv and reg

This is currently required as the argparsers for the conv and reg
modules are not yet integrated into the new registry architecture.
Instead, the workflow flags/values have to be manually remapped to the
module parsers. This can also be automated as all information is in the
objects. Eventually the default runner can be used once conv and reg
have been refactored away from hardcoded parser to object references.

jono3030 committed:
[fix] replace template commands with actual commands to call scripts

jono3030 committed:
[feat] changed manual UUID creation to default_factory

jono3030 committed:
[feat] remove manual UUID's from objects

jono3030 committed:
[feat] improve registry override flag/value methods

jono3030 committed:
[fix] fix runner not returning value if None

jono3030 committed:
[feat] add generate patch module to workflow

jono3030 committed:
[feat] add preprocessing parallel module to workflow

jono3030 committed:
[feat] add inference module to workflow

jono3030 committed:
[feat] add patch stacking module to workflow

jono3030 committed:
[feat] add raw data normalization module to workflow

jono3030 committed:
[feat] add CliGroup ENUMS for newly added modules

jono3030 committed:
[wip] add skeletonization module to workflow

This is a WIP as I ran into an issue with cucim-cu12, which is required
by the skeletonization module. cucim-cu12 does not support the version
of Python (3.7.3) installed in the MIRACL container. I have to think
about a workaround before the skeletonization module can be fully
implemented.

jono3030 committed:
[fix] create a separate Conda env for skeletonization, install required packages and call Python env bin directly from the runner

mapl3-env.yaml:
```
name: mapl3-env
channels:
  - defaults
dependencies:
  - python=3.10.8
  - pip
  - pip:
      - annotated-types==0.7.0
      - click==8.3.0
      - cucim-cu12==25.10.0
      - cupy-cuda12x==13.6.0
      - fastrlock==0.8.3
      - imageio==2.37.0
      - joblib==1.5.2
      - lazy_loader==0.4
      - MIRACL==2.4.0
      - networkx==3.4.2
      - numpy==2.2.6
      - packaging==25.0
      - pandas==2.3.3
      - pillow==12.0.0
      - pydantic==2.12.3
      - pydantic_core==2.41.4
      - python-dateutil==2.9.0.post0
      - pytz==2025.2
      - scikit-image==0.25.2
      - scipy==1.15.3
      - six==1.17.0
      - tifffile==2025.5.10
      - typing-inspection==0.4.2
      - typing_extensions==4.15.0
      - tzdata==2025.2
```

jono3030 committed:
[feat] add voxelization module to workflow and improve runner and template import methods

Runners and templates have separate import aggregators now that can be
imported to the main workflow script with '*'. Makes for a much cleaner
workflow interface.

jono3030 committed:
[feat] add more verbose information output to CLARITY Allen method

jono3030 committed:
[wip] temporarily prevent tiff to nii and warping from running

jono3030 committed:
[feat] change utility fns to __init__.py package

[feat] change utility fns to __init__.py package

jono3030 committed:
[feat] add warping module to MAPL3 workflow

jono3030 committed:
[feat] add feature extract module to MAPL3 workflow

jono3030 committed:
[feat] import utililty fns from package

jono3030 committed:
[feat] update runner and template facades

jono3030 committed:
[feat] split RAW input folder into separate signal and autoflor flags

jono3030 committed:
[fix] tiny fix for obsolete import

jono3030 committed:
[fix] expose patch stacking dtype arg to user in workflow

jono3030 committed:
[fix] type annotation fixes

jono3030 committed:
[feat] add module skip functionality

jono3030 committed:
[fix] add flag that should have been included in commit e6cf99ef

jono3030 committed:
[wip] set script to mock script

This eventually needs a better solution in the template registry.

jono3030 committed:
[feat] awesome refactor from imperative to declarative registry

configs/modules.yaml: Contains declarative config for modules incl. template and runner info.
registry.py: Now contains class method to return a list of all loaded modules for sanity checks.
registry_loader.py: Parses modules.yaml and registers modules to registry. Contains closure method to access runner data.
run_mapl3_workflow.py: WIP main() to run pipeline with registered modules.
runners/generic_runner.py: flag_map is optional. Also contains an execute flag now to selectively execute subcommand.

jono3030 committed:
[wip] test runs for pipeline - old imperative method

jono3030 committed:
[feat] add automatic mapping of flow flags to module flags to registry loader

Since each objects contains all information for workflow and modules
flags, it makes sense to automatically create flag mappings for
conversion. Explicit definitions in the YAML take precendent over
automated mappings. There is now also a validation to make sure that a
flag_map (can be an empty dict) is present in the YAML config. Lastly,
if the ModuleType is MODULE, the flags for the module will be used
directly instead of translating from the workflow maps as the assumption
is that the module is being run as a standalone module.

jono3030 committed:
[feat] update serializer

jono3030 committed:
[fix] convert list input to parser to joined string in generic runner

jono3030 committed:
[feat] add connectors and skips and update flag map for preprocessing_parallel

[feat] update workflow status for preprocessing_parallel

jono3030 committed:
[fix] fix incorrect object type in conversion module

jono3030 committed:
[feat] improve stdout vars overview, fix utilfns import and remove sys.exit() for production

jono3030 committed:
[fix] fix utilfns import and remove sys.exit() for production

jono3030 committed:
[fix] remove sys.exit() for production

jono3030 committed:
[fix] fix stdout summary formatting and remove exit 1 for production

jono3030 committed:
[fix] fix CLARITY Allen registration method

Restoring miracl_reg_clar-allen.sh and miracl_reg_clar-allen_utility.py from the dev branch. MAPL3 worklfow was failing for registration which will be fixed in this commit.

jono3030 committed:
[fix] fix argparser issue for None input and remove sys.exit() for production

[fix] remove sys.exit() for production

[feat] add method to correctly parse None input

jono3030 committed:
[fix] update module YAML with correct flag mappings

jono3030 committed:
[feat/fix] update GPU forwarding directives in docker-compose.yml

jono3030 committed:
[fix] fix inference objects attributes for workflow run

jono3030 committed:
[docs] merge latest docs and create first version of MAPL3 workflow tutorial

jono3030 committed:
[docs] improved MAPL3 docs and tutorials for submission
